src/Raven.Server/Documents/Indexes/Static/IndexCompiler.cs

```diff
@@ -224,87 +226,203 @@ private static CompilationResult CompileInternal(string originalName, string cSh
+                        continue;
+                    }
+
+                    if (additionalAssembly.AssemblyPath != null)
+                    {
+                        newReferences.Add(FromAssemblyPath(additionalAssembly.AssemblyPath));
+                        continue;
+                    }
+
+                    if (additionalAssembly.PackageName != null)
+                    {
+                        newReferences.AddRange(FromPackage(additionalAssembly.PackageName, additionalAssembly.PackageVersion, additionalAssembly.PackageSourceUrl));
+                        continue;
+                    }
+
+                    if (additionalAssembly.Usings != null && additionalAssembly.Usings.Count > 0)
+                        continue;
+
+                    throw new NotSupportedException($"Not supported additional assembly: {additionalAssembly}");
+                }
             }
-            for (int i = 0; i < knownManageRefs.Length; i++)
+
+            return newReferences.ToArray();
+
+            static MetadataReference FromAssemblyName(string name)
+            {
+                try
+                {
+                    var assemblyName = new AssemblyName(name);
+                    var assembly = Assembly.Load(assemblyName);
+                    return CreateMetadataReferenceFromAssembly(assembly);
+                }
+                catch (Exception e)
+                {
+                    throw new IndexCompilationException($"Cannot load assembly '{name}'.", e);
+                }
+            }
+
+            static MetadataReference FromAssemblyPath(string path)
             {
-                newReferences[i + References.Length] = knownManageRefs[i].Reference;
+                try
+                {
+                    var assembly = LoadAssembly(path);
+                    return CreateMetadataReferenceFromAssembly(assembly);
+                }
+                catch (Exception e)
+                {
+                    throw new IndexCompilationException($"Cannot load assembly from path '{path}'.", e);
+                }
+            }
+
+            static List<MetadataReference> FromPackage(string packageName, string packageVersion, string packageSourceUrl)
+            {
+                try
+                {
+                    packageSourceUrl ??= "https://api.nuget.org/v3/index.json";
+
+                    if (string.IsNullOrWhiteSpace(packageName))
+                        throw new ArgumentException($"'{nameof(packageName)}' cannot be null or whitespace", nameof(packageName));
+
+                    if (string.IsNullOrWhiteSpace(packageVersion))
+                        throw new ArgumentException($"'{nameof(packageVersion)}' cannot be null or whitespace", nameof(packageVersion));
+
+                    var paths = AsyncHelpers.RunSync(() => MultiSourceNuGetFetcher.Instance.DownloadAsync(packageName, packageVersion, packageSourceUrl));
```

Comment:
Do we need to worry about timeouts here? This can take a _while_, the ML package is > 100MB.

Reply:
Timeouts are controlled by the CTS (None). Do we want to introduce some or infinite is OK?

Reply:
My main issue is whatever the index creation is going to take a long enough time that we'll get a timeout on the request.
What do we do then?

Reply:
There is no timeout set on those requests IIRC?